### ISSUE TITLE
Possible to expand on bands extraction?

### ISSUE THREAD

**Issue by phke5716** [kvos/CoastSat#75]
**Is your feature request related to a problem? Please describe.**
No, not related to a problem! Just would like to learn more about the background processes :)

**Describe the solution you'd like**
Hi @kvos , by now (and my emails) you are well aware that I am a big fan of your academic pubs on the subject-content (methodology) for your coastsat module.  I was wondering, when reading both the larger pub (methodology deep) and shorter pub (explaining coastsat high-level process) if you would be able to append the processing of the imagery between the 'download' and before 'shoreline extraction'.

For example, during image processing, are the images downloaded inidividual bands, or, are those extracted separately during the image?

I will try to use this example 👍  :

From SDS_preprocess.preprocess_single_image(), there is the following line:

```
# read all bands
        data = gdal.Open(fn, gdal.GA_ReadOnly)
        georef = np.array(data.GetGeoTransform())
        bands = [data.GetRasterBand(k + 1).ReadAsArray() for k in range(data.RasterCount)]
        im_ms = np.stack(bands, 2)
```

However, in the source docs for arguments, SDS_classify.label_images(), you mention that fn = ( a list of filenames, one filename for each band"

```
# get individual RGB image
            im_RGB = SDS_preprocess.rescale_image_intensity(im_ms[:,:,[2,1,0]], cloud_mask, 99.9)
            im_NDVI = SDS_tools.nd_index(im_ms[:,:,3], im_ms[:,:,2], cloud_mask)
            im_NDWI = SDS_tools.nd_index(im_ms[:,:,3], im_ms[:,:,1], cloud_mask)
```

**Additional context**
Given the data above, I am wondering how the processing is actually carried out. For example, I have the following  understanding (in my small head):

Download Images (inputs, etc) [Let's pretend there is only one capture for the inputs, a L7 (B,G,R,NIR,SWIR1)] ---> Image Stored in (./data) ===skip ahead in process=== ---> Use pre-processing, gather the bands and stack RGB together, also extract im_NDVI & im_NDWI.

Obviously, from here - it differs based on creating new location (classifier training) or simply receiving the input. However, I would love to understand if we are pulling in a single image or if that image is being downloaded by band (i.e. are we getting ~7 copies of the image at different bands or a single raster)?

Thanks @kvos , stay awesome

**Comment by kvos** [comment]
hi @phke5716 ,
So each image is multi-spectral so it has several bands. In this case we only need RGB,NIR,SWIR1 and the QA band to run CoastSat, so the other bands are ignored (SWIR2, thermal..etc). All the bands with the same resolution are stacked together and downloaded as a .tif file. Then, the .tif file is read and preprocessed.

BTW I didn't receive your emails, most probably they ended up in the junk folder (that's why is better not to email these days but to use this forum)

**Comment by phke5716** [comment]
thanks for replying so quickly @kvos . I know you said the forum was best for posting, but I was hoping we could arrange a time to discuss coaststat more. Although I have read your literature on the tool (somewhat obsessively lol), I would love the chance to speak to more aspects of it, if possible. 

For example, I wanted to discuss the possibility of modifying your creation, changing certain aspects of the scripting, to expand the capability outside of EE. While I have done this somewhat successifuly, I have only done so in dev and would of course not publish (or commit) without discussing with you first. I hope you do have time, it would be wonderful to discuss further. Some general questions/comments are provided below for your considerations.

[Things I have modified, in DEV]
+Changes made to platform/vehicle : Expanded to allow a greater swath of sats, including DigitalGlobe platforms.

+Modified calls to ```retrieve metadata``` and others, to allow extraction of metadata from OTH platforms and file types.

Things for consultation/consideration:
-Cloud masking: As some platforms do not have a QA band, is there a way to omit cloud masking? From the scripting, this plays an important part - being present in much of the required arguments. I have adjusted to (try) and remove the requirement, but the cascading effects have made the process slow.

**Comment by kvos** [comment]
hi @phke5716 making CoastSat compatible with other satellite missions would be great!
You can omit cloud masking by just creating a boolean array with all False: 
- `cloud_mask= np.zeros((im_ms.shape[0],im_ms.shape[1)).astype(bool)`
You can also adjust the spectral indices using different bands to calculate NDVI and NDWI. NDVIThe spectral indices are mainly used for the classifier so you can train a classifier for your specific sensor with different spectral indices as the ones you mentioned, it should work fine as well. The only index that may be hard to replace is the MNDWI (SWIR1 - Green)/(SWIR1 + Grenn) as this is the one that is used to contour the shorelines. If you are working on a fork, I can also have a look and any pull request will be highly appreciated! We are currently adding compatibility with Planet Labs imagery, so also having Digital Globe would be a great enhancement.
Are you working for a universitiy?

**Comment by phke5716** [comment]
Hello @kvos, I am using the band combinations from Maglione, Parente &  Vallario (2014) [1] to calculate the indices {NDVI, NDWI}. 

Because the imagery is not accessed from EE (although, I have emailed them regarding this), there is some substantial changes to both the ‘get_metadata’ and ‘preprocess’.

Do you have a good email I can reach out to? That way I can expand in more detail ^_^

[1]https://www.tandfonline.com/doi/pdf/10.5721/EuJRS20144739”

**Comment by phke5716** [comment]
@kvos forgot to add; is there a reason you chose MLP over CNN? I was thinking, ArcGIS has recently added a ‘deep learning’ toolbox. From what I gather, the labeled raster is chipped and exported where it is trained (usually via CNN). I was thinking of doing something similar but wonder if you already approached this and decided to go a different route - btw, I believe (based off your writings) it was considered but pixel-wise classification was more accurate?

**Comment by phke5716** [comment]
@kvos see, omg this is why chat via email (or tele) would work best-my brain and fingers don’t match ^_^. Regarding Arc, I have tried (unsuccessfully) to train a new classifier from a labeled raster. Everything went ok until trying to prepare the data (after labeling). I can show you via fork later today. Short story: raster is read in and pixel values are the labels [0,1,2,3]. However, joining them with the coasat training data is proving elusive- mostly due to the dictionary, specifically when ‘bool’ is checked.

**Comment by kvos** [comment]
There is a chat https://gitter.im/CoastSat/community, it's also on one of the badges in the README 

CNN is great and state-of-the-art (pixel-wise MLP quite obsolete) but requires much more training data - and keep in mind that to train a CNN you have to label the entire image, which is not the case in the coastsat training data. 

In my opinion, if you want to go the CNN way the best would be transfer learning, and finding a good pretrained CNN for semantic classification of satellite images

**Comment by phke5716** [comment]
Whoop! Thought I closed this ^_^
